src/clib/vector.d

```diff
@@ -38,32 +38,144 @@ struct vector(T, A: IAllocator!T = allocator!T) if (!is(T == bool)) {
     @property size_t size() @nogc nothrow { return _size; }
 
     /// Returns true if vector is empty, i.e `size == 0`
-    @property bool empty() @nogc nothrow { return _size == 0; }
+    @property bool empty() const @nogc nothrow { return _size == 0; }
 
     /// Returns last element or `T.init` if `size == 0`
     @property T back() @nogc nothrow { return _size == 0 ? T.init : _data[_size - 1]; }
 
     /// Returns first element or `T.init` if `size == 0`
     @property T front() @nogc nothrow { return _size == 0 ? T.init : _data[0]; }
 
-    /// Returns iterator to beginning
-    @property auto begin() @nogc nothrow {
-        return forward_iterator!(vector!(T, A), T)(this, _size);
+    VectorIterator!(T) begin() @nogc nothrow {
+        return VectorIterator!(T)(&this, false);
     }
 
-    /// Returns iterator to end
-    @property auto end() @nogc nothrow {
-        return forward_iterator!(vector!(T, A), T)(this, 0);
+    VectorIterator!(T) end() @nogc nothrow {
+        return VectorIterator!(T)(&this, true);
+    }
+
+    ConstVectorIterator!(T) begin() const @nogc nothrow {
+        return ConstVectorIterator!(T)(&this, false);
+    }
+
+    ConstVectorIterator!(T) end() const @nogc nothrow {
+        return ConstVectorIterator!(T)(&this, true);
     }
 
     /// Returns backward iterator to beginning (for reverse iteration)
-    @property auto rbegin() @nogc nothrow {
-        return backward_iterator!(vector!(T, A), T)(this, _size);
+    ReverseVectorIterator!(T) rbegin() @nogc nothrow {
+        return ReverseVectorIterator!(T)(&this, false);
     }
 
     /// Returns backward iterator to end (for reverse iteration)
-    @property auto rend() @nogc nothrow {
-        return backward_iterator!(vector!(T, A), T)(this, 0);
+    ReverseVectorIterator!(T) rend() @nogc nothrow {
+        return ReverseVectorIterator!(T)(&this, true);
+    }
+
+    /// Returns const backward iterator to beginning (for reverse iteration)
+    ConstReverseVectorIterator!(T) rbegin() const @nogc nothrow {
+        return ConstReverseVectorIterator!(T)(&this, false);
+    }
+
+    /// Returns const backward iterator to end (for reverse iteration)
+    ConstReverseVectorIterator!(T) rend() const @nogc nothrow {
+        return ConstReverseVectorIterator!(T)(&this, true);
+    }
+
+    private struct VectorIterator(T) {
```

Comment:
I'd say prefer using iterators from clib.iterator and avoid declaring private types that would be returned from functions